File: express-ts-bun-backend/tests/server-lifecycle.test.ts
import type { Logger } from 'pino';
import { describe, expect, it, vi } from 'vitest';
import { createServerLifecycle } from '../src/server-lifecycle';

function createFakeLogger() {
  return {
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
  } as unknown as Logger;
}

describe('server lifecycle', () => {
  it('calls graceful close and exits with 0 on success', async () => {
    const fakeServer = {
      close: (callback: (error?: Error) => void) => {
        callback();
      },
    };

    const onBeforeShutdown = vi.fn();
    const onExit = vi.fn();
    const lifecycle = createServerLifecycle(fakeServer, {
      timeoutMs: 25,
      logger: createFakeLogger(),
      onBeforeShutdown,
      onExit,
    });

    await lifecycle.shutdown('SIGTERM', 0);

    expect(onBeforeShutdown).toHaveBeenCalledTimes(1);
    expect(onExit).toHaveBeenCalledWith(0);
    expect(lifecycle.isShutdownInProgress()).toBe(true);
  });

  it('times out and exits with 1 when close does not finish', async () => {
    const fakeServer = {
      close: (_callback: (error?: Error) => void) => {
        // Intentionally never calling the callback to simulate a hung shutdown.
      },
    };

    const onExit = vi.fn();
    const lifecycle = createServerLifecycle(fakeServer, {
      timeoutMs: 10,
      logger: createFakeLogger(),
      onExit,
    });

    await lifecycle.shutdown('SIGINT', 0);

    expect(onExit).toHaveBeenCalledWith(1);
  });
});

File: flask-python-backend/tests/server_lifecycle_test.py
import asyncio
import logging

import pytest

from src.server_lifecycle import create_server_lifecycle


class SuccessfulServer:
    async def close(self) -> None:
        await asyncio.sleep(0)


class HungServer:
    async def close(self) -> None:
        await asyncio.sleep(60)


def _logger() -> logging.Logger:
    logger = logging.getLogger('server-lifecycle-test')
    if not logger.handlers:
        logger.addHandler(logging.NullHandler())
    return logger


@pytest.mark.asyncio
async def test_calls_graceful_close_and_exits_with_0_on_success() -> None:
    exit_codes: list[int] = []

    lifecycle = create_server_lifecycle(
        SuccessfulServer(),
        timeout_ms=25,
        logger=_logger(),
        on_exit=exit_codes.append,
    )

    await lifecycle.shutdown('SIGTERM', 0)

    assert exit_codes == [0]
    assert lifecycle.is_shutdown_in_progress() is True


@pytest.mark.asyncio
async def test_times_out_and_exits_with_1_when_close_hangs() -> None:
    exit_codes: list[int] = []

    lifecycle = create_server_lifecycle(
        HungServer(),
        timeout_ms=10,
        logger=_logger(),
        on_exit=exit_codes.append,
    )

    await lifecycle.shutdown('SIGINT', 0)

    assert exit_codes == [1]
